### DISCUSSION TITLE
MUST

### DISCUSSION THREAD

**Discussion by jovanHillion** [Sinan-Karakaya/R-Type#15] [Mandatory]
# Project

- [x] The project MUST be fully self-contained regarding its dependencies and 3rd-party libraries that may be used.
- [x] The project can be built and run without altering anything on the system.
- [x] It does not rely on system-wide pre-installed libraries or development headers (for example, installed manually with apt install or dnf install), except for the standard C++ compilers & libraries and some obvious low-level
and platform-specific graphical or sound system libraries such as OpenGL/X11/etc.
- [x] In particular, usage of SFML, SDL or any other kind of high-level 3rd-party library MUST be properly packaged along with the project, both at compilation time and runtime.
- [x] The project MUST use a proper method for handling 3rd-party dependencies, that could be ONE of the following:
	- Git submodules
	- CMake Fetch_Package or External_Project functionalities
	- A dedicated package manager : Conan or Vcpkg
- [ ] The evaluator should be able to build and run your project “Out of the Box”, only using its locally installed C++ compiler and environment, the CMake tool, and a shell.
- [x] The project MUST be Cross Platform (It has to build and run on Windows using Microsoft Visual C++ compiler AND Linux using GCC.)
- [x] You MUST provide a mean to build tarballs or installers for the game. Additionally, ready-made tarballs or
installers can be made available on the project online site.
- [x] You MUST write the documentation in English
- [x] You MUST have a readME !
- [x] It has to be short, nice, practical, and useful, with typical information such as project purpose, dependencies/requirements/supported platforms, build and usage instructions, license, authors/contacts, useful links or quick-start information, and so on.
- [ ] Developer Documentation: Architectural diagrams, Main systems overviews and description, and how this materializes in the code, Tutorials and How-To’s. Contribution guidelines and coding conventions are very useful too
- [ ] You MUST produce documentation at a higher level than only just classes/functions descriptions !
- [ ] The documentation have to be available and accessible in a modern way
- [x] Protocol Documentation

## Server
- [x] Must be a networked game
- [x] Server game must be Multithreaded

## Game Engine
- [x] MUST use CMake as its build system. No Makefiles are allowed !
- [x] You must demonstrate good usage of feature branches, merge requests, issues, tags for important milestones, commits content and description, etc.

# Part 1 Must

- [x] The game MUST be playable at the end of this part: with a nice star-field in background, players spaceships confront waves of enemy Bydos, everyone shooting missiles to try to get down the opponent.
- [x] The game MUST be a networked game: each player use a distinct Client program on the network,
connecting to a central Server having final authority on what is happening in the game.
- [x] The game MUST demonstrate the foundations of a game engine, with at least visible and decoupled
subsystems/layers for Rendering, Networking, and Game Logic.

## Server

- [x] The server MUST notify each client when a monster spawns, moves, is destroyed, fires, kills a player,
and so on.. ., as well as notifies others clients’ actions (a player moves, shoots, etc.).
- [x] The server MUST be multithreaded, or at least, MUST NOT block or wait for clients messages, as the
game must run frame after frame on the server regardless of clients’ actions!
- [x] If a client crashes for any reason, the server MUST continue to work and MUST notify other clients
in the same game that a client crashed. More generally, the server must be robust and be able to run
regardless of what’s wrong might happen.
- [x]  You MAY use the Asio library for networking, or rely on OS-specific network layer with an appropriate
encapsulation.

Note that your abstractions’ quality will be evaluated during the final defense, so pay close attention to
them.

## Client

- [x] It MUST contain anything necessary to display the game and handle player input, while everything related
to gameplay logic shall occur on the server. (This is particularly true for any kind of effect that have a great impact on gameplay (death of an enemy/player, pickup of an
item, etc.): all players have to play the same game, whose rules are driven by the server.

## Protocol

- [x] You MUST design a binary protocol for client/server communications.
- [x] You MUST use UDP for communications between the server and the clients. A second connection using
TCP can be tolerated, but you MUST provide a strong justification. In any event, ALL in-game communications MUST use UDP.
- [x] Malformed messages or packets MUST NOT lead the client or server to crash or consume
excessive memory.
- [x] You MUST document your protocol.

## Game Engine

- [x] Your game engine MUST be modular.

## Gameplay

- [ ] The client MUST display a slow horizontal scrolling background representing space with stars, planets.. .
This is the star-field.
- [ ] The star-field scrolling, entities behavior, and overall time flow must NOT be tied to the CPU speed. Instead, you MUST use timers.
- [x] Players MUST be able to move using the arrow keys.
- [ ] There MUST be Bydos slaves in your game, as well as missiles.
- [x] Monsters MUST be able to spawn randomly on the right of the screen.
- [x] The four players in a game MUST be distinctly identifiable (via color, sprite, etc.)

# LISTE DE COMPÉTENCE

## BLOC 2 : Concevoir une architecture logicielle

- [x] Mettre en place une veille légale et réglementaire prenant en compte les besoins des PSH en menant des recherches fréquentes sur les usages liés aux technologies, en recensant les aspects légaux et réglementaires parus et en participant régulièrement aux rencontres de la communauté professionnelle afin de minimiser la dette technique pour les aspects couverts par le projet.
- [x] Réviser régulièrement les protocoles existants, notamment au regard des nouvelles failles de sécurité identifiées afin de contribuer à l'utilisation de standards technologiques élevés au sein de l'entreprise, en impulsant une application régulière des nouveaux usages et outils à l'entreprise.
- [x] Présenter une solution technique créative, en collaboration avec l'équipe projet et ses différentes expertises, en prenant en considération les différentes contraintes apportées par le client (économique, RSE, …) ou imposées par l'environnement technique dans le but de résoudre la problématique exposée.
- [ ] Sélectionner une hypothèse d'architecture et l'urbanisme de la solution logicielle ou web, afin de garantir l'intégration et la pérennité d'une solution, en prenant en compte le reste de l'écosystème technique présent au sein de l'entreprise ainsi que l'accessibilité numérique de la solution.
- [x] Traduire les spécifications techniques et fonctionnelles en un système cohérent de composants logiciels en mobilisant son expertise et en s'appuyant sur l'état de l'art en termes d'architecture logicielle afin de produire une solution technique adaptée au besoin du client.
- [x] Segmenter chaque problème complexe en un ensemble de sous-problèmes afin d'obtenir des tâches atomiques dans un objectif de performance, d'adaptabilité et de maintenabilité en fonction des besoins du client.
- [x] Identifier des solutions existantes ou originales afin de répondre à chaque problème posé en tenant compte des contraintes de performance et de scalabilité de la solution et de son environnement d'exécution.
- [ ] Sélectionner les solutions de persistance de données (fichier texte ou binaire, format de fichier structuré, base de donnée, etc.) en s'appuyant sur son expertise et celle de l'équipe projet et en mobilisant l'état de l'art afin de s'adapter aux contraintes techniques, fonctionnelles et de sécurité de l'application en terme de stockage de données.
- [x] Sélectionner les structures de données répondant aux contraintes de l'application en tenant compte de leur complexité algorithmique et spatiale (tableaux, listes, sets, tables de hashage...) dans un objectif de performance, de maintenabilité et d'évolutivité de l'application.

## BLOC 3 : Créer une architecture logicielle pour des applications dites “lourdes”

- [ ] Concevoir les interfaces (GUI, TUI, CLI, …) afin de garantir une bonne expérience utilisateur dans le respect des conventions d'UI et d'UX spécifiques aux modalités d'interactions concernées, en optimisant les vues et en respectant les critères reconnus d'accessibilité.
- [ ] Vérifier l'intégrité des données traitées en s'appuyant sur les techniques de vérification de données et dans le respect des normes de conformité et de sécurité afin de s'assurer que la donnée n'a pas été corrompue et de prévenir ainsi les dysfonctionnements du logiciel.
- [x] Sélectionner les solutions techniques adaptées (protocoles, formats de fichiers...) en utilisant des composants logiciels tiers (bibliothèque) afin de proposer des implémentations permettant l'interopérabilité avec d'autres systèmes. 
- [x] Rédiger le code à l'aide du langage informatique adapté au logiciel en implémentant les solutions techniques précédemment identifiées, afin de concrétiser la vision et la valeur du produit par le client.   
- [x] Intégrer l'usage de codes tiers au code produit en s'appuyant sur la documentation et en suivant les instructions relatives aux solutions retenues afin d'optimiser le temps de production et l'efficacité ainsi que la maintenabilité du code.

**Comment by mlbonniec** [comment]
Récapitulatif follow-up 18/12/2023

- [ ] Un engine ?
- [ ] Architecture custom ?
- [ ] ECS / OOP ?
- [ ] Gestion des dépendances ? Abstractions etc.
- [ ] Entités mises à jour de manière générique ? Bootstrap fait correctement
- [ ] Player / NPC peuvent ils être contrôlés, par input (avec IA, ou IF PLAYER vs NPC)
- [ ] Configurable ?
- [ ] Pas de réseau = pas de crédit
- [ ] Qui a l’avis final sur la position ? Serveur, client… ?
- [ ] Consistant ? Tous les clients sont synchronisés ? Smooth / Téléportation
- [ ] Rollback nécessaire ? Peut-être pas sur un R-Type
- [ ] TCP ou UDP ? Pourquoi l’un plutôt que l’autre
- [ ] Gestion d’erreur ?
- [ ] Gestion des erreurs dans le protocole ? Que se passe-t-il si le packet est invalide ? À INDIQUER DANS LA DOCUMENTATION RFC
    - [ ] In case of error, nothing is returned
- [ ] ECS & Réseau fonctionnel en priorité
- [ ] Le jeu passe après, même si cela fait perdre des points
- [ ] Le projet build avec CMake ? Linux & Windows ?
- [ ] Projet self contained ? Ou dépendances
- [ ] README incomplet -> Ajouter les autre README dans le README principal
- [ ] Documentation protocol réseau disponible ?
- [ ] Documentation développeur utilisable ? (ex. Doxygen, si envie d’ajouter des choses dans le code)
    - [ ] Clair, illustré
    - [ ] Diagramme d’architecture pour le fonctionnement
    - [ ] Documentation du fonctionnement de l’API, de l’ajout des systèmes, scripting etc.
    - [ ] Préférence pour des documentation lisibles (ex. Github Wiki, README etc.)
- [ ] Tutoriel / Documentation scripting
- [ ] Librairies encapsulés (ça vaut des points)
- [ ] Engine construit en modules ?
- [ ] Engine configurable ? (ex. désactiver la gravité depuis l’engine) 
- [ ] Chargement dynamique de parties de l’engine ?
